### ISSUE TITLE
HttpClientHandler.ServerCertificateCustomValidationCallback not called in dockerized AspNet.Core application

### ISSUE THREAD

**Issue by TorbenJ** [dotnet/runtime#1931]
I have an AspNet.Core 3.1 application that needs to do some HTTPS requests against an internal service. For this I'm using standard `HttpRequestMessage`s, nothing special.  
If I run the application using docker-compose (using Docker CE on Windows 10) these requests always throw exceptions: 
```
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception.
 ---> Interop+OpenSsl+SslException: SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
 ---> Interop+Crypto+OpenSslCryptographicException: error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol
   --- End of inner exception stack trace ---
   at Interop.OpenSsl.DoSslHandshake(SafeSslHandle context, Byte[] recvBuf, Int32 recvOffset, Int32 recvCount, Byte[]& sendBuf, Int32& sendCount)
   at System.Net.Security.SslStreamPal.HandshakeInternal(SafeFreeCredentials credential, SafeDeleteContext& context, ArraySegment`1 inputBuffer, Byte[]& outputBuffer, SslAuthenticationOptions sslAuthenticationOptions)
   --- End of inner exception stack trace ---
   at System.Net.Security.SslStream.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.PartialFrameCallback(AsyncProtocolRequest asyncRequest)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Security.SslStream.ThrowIfExceptional()
   at System.Net.Security.SslStream.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslStream.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__65_1(IAsyncResult iar)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean allowHttp2, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at WebApplication1.Controllers.FooController.Get() in C:\Users\tjonas\source\repos\ConsoleApp4\WebApplication1\Controllers\FooController.cs:line 37
```

I then tried to just bypass SSL validation by passing a custom `HttpClientHandler` to my `HttpClient`:  
```
var httpClientHandler = new HttpClientHandler
{
    ServerCertificateCustomValidationCallback = (s, c, ch, e) =>
    {
        return true;
    },
};

var httpClient = new HttpClient(httpClientHandler);
```

This didn't seem to have any effect though. After long googling around without any success I decided to just put a breakpoint inside the callback.  
I noticed that the callback isn't getting called at all when running in Docker. Without using Docker it gets called as expected.

So in short what is working right now:  
- requesting data from internal and external services using HTTPS while server runs standalone i.e. without docker
- requesting data from external services using HTTPS while running server in Docker

What doesn't work right now:
- requesting data from internal services using HTTPS while running server in Docker

I'm really clueless by now, this issue didn't occur back in AspNet.Core 2.0 iirc and I couldn't find any hint on why this happens.  

My docker setup running on Windows 10 1809 17763.973 x64:  
![grafik](https://user-images.githubusercontent.com/9077102/72715284-7bf12a80-3b70-11ea-9f1c-294797741804.png)

Hope anyone of you knows where to look next.

**Comment by TorbenJ** [comment]
Posted in the wrong repo, sry

**Comment by scalablecory** [comment]
Are your client and server using incompatible TLS versions? "error:1425F102:SSL routines:ssl_choose_client_version:unsupported protocol"

**Comment by TorbenJ** [comment]
Reopening because AspNet.Core Repo recommended it.  
@scalablecory I don't have any client, just the server application.  
If I run the application without Docker and request data from a different endpoint (e.g. `https://httpbin.org/get`) it works fine. So it seems to be an issue when using Docker + requesting from our internal services.

**Comment by TorbenJ** [comment]
I also updated the issue description and listed things that work and things that don't

**Comment by wfurt** [comment]
Do you have runable repro @tobrej ? From the the description it seems like the container does not have same certificates and trust as the machine it self. 
However, you _should_ be able to override that with custom validation and we should be be able to trace why that does not work for you.

**Comment by karelz** [comment]
No answer for 3 weeks, closing. If there is more info, please feel free to reopen.

**Comment by TorbenJ** [comment]
I'm sorry for not answering, I resolved the issue in the meantime.

There was nothing wrong with the application or Docker itself, it was a combination of the internal service and the debian version used by the ASP.NET Core docker image.
Debian blocks SSL/TLS versions <= 1.0 if I remember correctly and the internal service's apache wasn't updated for some time and used TLS 1.0.  

The solution is to either reconfigure the image to allow 1.0 or, in our case, to update the apache to use a newer TLS version.
